commerce-test/src/testIntegration/java/com/liferay/commerce/test/CPRuleTest.java

```diff
@@ -0,0 +1,222 @@
+import com.liferay.commerce.product.test.util.CPTestUtil;
+import com.liferay.commerce.product.util.CPDefinitionHelper;
+import com.liferay.portal.kernel.dao.orm.QueryUtil;
+import com.liferay.portal.kernel.model.Group;
+import com.liferay.portal.kernel.model.User;
+import com.liferay.portal.kernel.search.SearchContext;
+import com.liferay.portal.kernel.security.auth.PrincipalException;
+import com.liferay.portal.kernel.security.permission.PermissionChecker;
+import com.liferay.portal.kernel.security.permission.PermissionCheckerFactoryUtil;
+import com.liferay.portal.kernel.security.permission.PermissionThreadLocal;
+import com.liferay.portal.kernel.test.rule.AggregateTestRule;
+import com.liferay.portal.kernel.test.rule.DeleteAfterTestRun;
+import com.liferay.portal.kernel.test.util.GroupTestUtil;
+import com.liferay.portal.kernel.test.util.UserTestUtil;
+import com.liferay.portal.kernel.workflow.WorkflowConstants;
+import com.liferay.portal.test.rule.Inject;
+import com.liferay.portal.test.rule.LiferayIntegrationTestRule;
+import com.liferay.portal.test.rule.PermissionCheckerTestRule;
+
+import java.util.List;
+
+import org.frutilla.FrutillaRule;
+
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+
+/**
+ * @author Luca Pellizzon
+ */
+@RunWith(Arquillian.class)
+public class CPRuleTest {
+
+	@ClassRule
+	@Rule
+	public static final AggregateTestRule aggregateTestRule =
+		new AggregateTestRule(
+			new LiferayIntegrationTestRule(),
+			PermissionCheckerTestRule.INSTANCE);
+
+	@Before
+	public void setUp() throws Exception {
+		_group = GroupTestUtil.addGroup();
+
+		_user1 = UserTestUtil.addUser();
+
+		_user2 = UserTestUtil.addUser();
+
+		PermissionChecker permissionChecker =
+			PermissionCheckerFactoryUtil.create(_user1);
+
+		PermissionThreadLocal.setPermissionChecker(permissionChecker);
+	}
+
+	@Test
+	public void testUserSegmentCategoryCPDefinitionHelper() throws Exception {
+		frutillaRule.scenario(
+			"Search for CatalogEntries accessible from a given user"
+		).given(
+			"A user, a category with a product"
+		).when(
+			"I try to get the CatalogEntries that the user can view"
+		).then(
+			"The results will contain only accessible element from that " +
+				"specific user"
+		);
+
+		CPRule cpRule = CPRuleTestUtil.addUserSegmentCPRule(
+			_group.getGroupId(), _user1.getUserId());
+
+		AssetCategory assetCategory = CPRuleTestUtil.addAssetCategoryToCPRule(
+			cpRule);
+
+		CPRuleTestUtil.setCPRulesThreadLocal(
+			_group.getGroupId(), 0, _user1.getUserId());
+
+		CPInstance cpInstance = CPTestUtil.addCPInstance(_group.getGroupId());
+
+		CPTestUtil.addCPInstance(_group.getGroupId());
+		CPTestUtil.addCPInstance(_group.getGroupId());
+
+		CPDefinition expectedCPDefinition = cpInstance.getCPDefinition();
+
+		CPRuleTestUtil.addCategoryToCPDefinition(
+			expectedCPDefinition, assetCategory.getCategoryId());
```

Comment:
Ideally these lines

    CPRule cpRule = CPRuleTestUtil.addUserSegmentCPRule(_group.getGroupId(), _user1.getUserId());

    AssetCategory assetCategory = CPRuleTestUtil.addAssetCategoryToCPRule(cpRule);

should come after this line

    CPDefinition expectedCPDefinition = cpInstance.getCPDefinition();

because of the order the variables are used in.